designs/http_client_design.md

```diff
@@ -0,0 +1,250 @@
+* Invent their own patterns for how FAIR and retry interact.
+
+This leads to:
+
+* **Inconsistent integrations**: everyone wires it differently.
+* **Higher friction to adoption**: more boilerplate than “just use FAIR”.
+
+We want a standard, idiomatic way to “augment” HTTP clients with FAIR now, and a pattern we can generalize to other client types later (gRPC, WebSockets, etc.).
+
+## Goals / Non-Goals
+
+**Goals**
+
+* Provide a **drop-in wrapper** (`FairRoundTripper`) for existing `http.Client`s.
+* Provide a **standalone client** (`FairHTTPClient`) with production defaults.
+* Make FAIR integration **idiomatic Go** (RoundTripper-based, composable).
+* Handle **client ID extraction** and **FAIR outcome reporting** automatically.
+* Include **basic resiliency**: retry with exponential backoff + idempotency awareness.
+* Keep **overhead low** (target <5% P99 latency increase vs raw `http.Client` for non-throttled requests).
+
+**Non-Goals / Out of Scope (for this design)**
+
+* Server-side middleware changes.
+* gRPC, WebSocket, or other non-HTTP transports (future designs).
+* Circuit breakers, metrics/tracing, request hedging, client-side load balancing (future work).
+
+## Proposed Design
+
+### Package & Types
+
+New package: `pkg/fairclient`.
+
+Core types (high-level):
+
+```go
+// FairRoundTripper wraps an http.RoundTripper with FAIR.
+type FairRoundTripper struct {
+    config: *FairRoundTripperConfig
+    transport: http.RoundTripper
+}
+
+// Config for FairRoundTripper.
+type FairRoundTripperConfig struct {
+    Tracker           *tracker.FairnessTracker          // required
+    ClientIDExtractor ClientIDExtractor                 // optional, default header-based
+    FallbackClientID  []byte                            // optional
+    OnThrottle        func(*http.Request, *request.RegisterRequestResult)
+    ThrottleStatusCode int                              // default 429
+    RetryConfig       *RetryConfig                      // optional
+}
+
+// Production-ready client with FAIR built in.
+type FairHTTPClient struct {
+    client *http.Client
+}
+
+// Configures FairHTTPClient.
+type FairHTTPClientConfig struct {
+    FairConfig      *FairRoundTripperConfig // required
+    Timeout         time.Duration           // default 30s
+    TransportConfig *TransportConfig        // connection pooling defaults
+}
+```
+
+Additional helpers:
+
+* `type ClientIDExtractor func(*http.Request) ([]byte, error)`
+* `HeaderClientIDExtractor(headerName string) ClientIDExtractor`
+* `ContextClientIDExtractor(key any) ClientIDExtractor`
+* `HashedClientIDExtractor(inner ClientIDExtractor) ClientIDExtractor`
+
+### Behavior
+
+**Request flow**
+
+On every request:
+
+1. **Extract client ID** with `ClientIDExtractor` (header by default, e.g. `X-Client-ID`).
+
+   * If extraction fails and `FallbackClientID` is set, use fallback.
+   * If extraction fails and no fallback, return an error immediately.
```

Comment:
Maybe instead of an error, make the default work?  What I meant was to use something like a empty-string or some pre-canned string - that will make everyone being rate-limited as from a client?